src/k8s/pkg/k8sd/api/certs_refresh.go

```diff
@@ -0,0 +1,263 @@
+		return response.InternalError(fmt.Errorf("missing CA certificates"))
+	}
+
+	certificates.CACert = *clusterConfig.Certificates.CACert
+	certificates.ClientCACert = *clusterConfig.Certificates.ClientCACert
+
+	g, ctx := errgroup.WithContext(r.Context())
+
+	for _, csr := range []struct {
+		name         string
+		commonName   string
+		organization []string
+		usages       []v1.KeyUsage
+		hostnames    []string
+		ips          []net.IP
+		signerName   string
+		certificate  *string
+		key          *string
+	}{
+		{
+			name:         fmt.Sprintf("k8sd-%d-worker-kubelet-serving", req.Seed),
+			commonName:   fmt.Sprintf("system:node:%s", snap.Hostname()),
+			organization: []string{"system:nodes"},
+			usages:       []v1.KeyUsage{v1.UsageDigitalSignature, v1.UsageKeyEncipherment, v1.UsageServerAuth},
+			hostnames:    []string{snap.Hostname()},
+			ips:          []net.IP{net.ParseIP(s.Address().Hostname())},
+			signerName:   "k8sd.io/kubelet-serving",
+			certificate:  &certificates.KubeletCert,
+			key:          &certificates.KubeletKey,
+		},
+		{
+			name:         fmt.Sprintf("k8sd-%d-worker-kubelet-client", req.Seed),
+			commonName:   fmt.Sprintf("system:node:%s", snap.Hostname()),
+			organization: []string{"system:nodes"},
+			usages:       []v1.KeyUsage{v1.UsageDigitalSignature, v1.UsageKeyEncipherment, v1.UsageClientAuth},
+			signerName:   "k8sd.io/kubelet-client",
+			certificate:  &certificates.KubeletClientCert,
+			key:          &certificates.KubeletClientKey,
+		},
+		{
+			name:        fmt.Sprintf("k8sd-%d-worker-kube-proxy-client", req.Seed),
+			commonName:  "system:kube-proxy",
+			usages:      []v1.KeyUsage{v1.UsageDigitalSignature, v1.UsageKeyEncipherment, v1.UsageClientAuth},
+			signerName:  "k8sd.io/kube-proxy-client",
+			certificate: &certificates.KubeProxyClientCert,
+			key:         &certificates.KubeProxyClientKey,
+		},
+	} {
+		csr := csr
+		g.Go(func() error {
+			csrPEM, keyPEM, err := pkiutil.GenerateCSR(
+				pkix.Name{
+					CommonName:   csr.commonName,
+					Organization: csr.organization,
+				},
+				2048,
+				nil,
+				csr.hostnames,
+				csr.ips,
+			)
+			if err != nil {
+				return fmt.Errorf("failed to generate CSR for %s: %w", csr.name, err)
+			}
+
+			if _, err = client.CertificatesV1().CertificateSigningRequests().Create(ctx, &certificatesv1.CertificateSigningRequest{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: csr.name,
+				},
+				Spec: certificatesv1.CertificateSigningRequestSpec{
+					Request:    []byte(csrPEM),
+					Usages:     csr.usages,
+					SignerName: csr.signerName,
+				},
+			}, metav1.CreateOptions{}); err != nil {
+				return fmt.Errorf("failed to create CSR for %s: %w", csr.name, err)
+			}
+			watcher, err := client.CertificatesV1().CertificateSigningRequests().Watch(ctx, metav1.SingleObject(metav1.ObjectMeta{Name: csr.name}))
+			if err != nil {
+				return fmt.Errorf("failed to watch CSR %s: %w", csr.name, err)
+			}
+
+			defer watcher.Stop()
```

Comment:
be careful, as this might fail because of transient errors like an apiserver restarting. this will presently fail, perhaps we would like to retry the watch until the context is cancelled.

It's OK as is, this is just an infrequent error path that we would like to avoid

Reply:
Should I introduce a more robust mechanism, such as wait.Until or wait.ExponentialBackoff for this wait? Or is a big enclosing for loop enough? Also, is it required that we retry the watcher creation even inside the watcher loop?

Reply:
When a watcher fails, you need to create a new one. This is not necessarily a requirement, but would give some robustness to the code.

See what we do in WatchConfigMap for this.